crates/wysiwyg/src/tests/test_to_markdown.rs

```diff
@@ -0,0 +1,228 @@
+    assert_to_md_no_roundtrip(
+        "abc<strong> def </strong>ghi",
+        "abc__ def __ghi",
+    );
+    assert_to_md(
+        "abc <strong>line1<br />line2<br /><br />line3</strong> def",
+        r#"abc __line1\
+line2\
+\
+line3__ def"#,
+    );
+}
+
+#[test]
+fn text_with_italic_and_bold() {
+    assert_to_md("<em><strong>abc</strong></em>", "*__abc__*");
+    assert_to_md("<em>abc <strong>def</strong></em> ghi", "*abc __def__* ghi");
+    assert_to_md(
+        "abc <em><strong>line1<br />line2</strong> def</em>",
+        r#"abc *__line1\
+line2__ def*"#,
+    );
+}
+
+#[test]
+fn text_with_strikethrough() {
+    assert_to_md("<del>abc</del>", "~~abc~~");
+    assert_to_md_no_roundtrip("abc<del>def</del>ghi", "abc~~def~~ghi");
+    assert_to_md("abc <del>def</del> ghi", "abc ~~def~~ ghi");
+    assert_to_md_no_roundtrip("abc<del> def </del>ghi", "abc~~ def ~~ghi");
+    assert_to_md(
+        "abc <del>line1<br />line2<br /><br />line3</del> def",
+        r#"abc ~~line1\
+line2\
+\
+line3~~ def"#,
+    );
+}
+
+#[test]
+fn text_with_underline() {
+    assert_to_md_no_roundtrip("<u>abc</u>", "abc");
+}
+
+#[test]
+fn text_with_inline_code() {
+    assert_to_md("<code>abc</code>", "`` abc ``");
+    // Inline code with a backtick inside.
+    assert_to_md("<code>abc ` def</code>", "`` abc ` def ``");
+    // Inline code with a backtick at the start.
+    assert_to_md("<code>`abc</code>", "`` `abc ``");
+    assert_to_md("abc <code>def</code> ghi", "abc `` def `` ghi");
+    assert_to_md("abc<code> def </code>ghi", "abc``  def  ``ghi");
+    // It's impossible to get a line break inside an inline code with Markdown.
+    assert_to_md_no_roundtrip(
+        "abc <code>line1<br />line2<br /><br />line3</code> def",
+        "abc `` line1 line2  line3 `` def",
+    );
+    assert_to_md_no_roundtrip(
+        "abc <code>def <strong>ghi</strong> jkl</code> mno",
+        "abc `` def __ghi__ jkl `` mno",
+    );
+}
+
+#[test]
+fn link() {
+    assert_to_md(r#"<a href="url">abc</a>"#, "[abc](<url>)");
+    assert_to_md_no_roundtrip(r#"<a href="u<rl">abc</a>"#, r#"[abc](<u\<rl>)"#);
+    // Empty link.
+    assert_to_md(r#"<a href="">abc</a>"#, r#"[abc](<>)"#);
+    // Formatting inside link.
+    assert_to_md(
+        r#"<a href="url">abc <strong>def</strong> ghi</a>"#,
+        r#"[abc __def__ ghi](<url>)"#,
+    );
+    assert_to_md(r#"<a href="(url)">abc</a>"#, r#"[abc](<\(url\)>)"#);
+}
+
+#[test]
+fn list_unordered() {
+    assert_to_md(
+        r#"<ul><li>item1</li><li>item2</li></ul>"#,
+        r#"* item1
+* item2"#,
+    );
+
+    assert_to_md_no_roundtrip(
+        r#"<ul><li>item1<ul><li>subitem1</li><li>subitem2</li></ul></li><li>item2</li></ul>"#,
+        r#"* item1
+  * subitem1
+  * subitem2
+* item2"#,
+    );
+}
+
+#[test]
+fn list_ordered() {
+    assert_to_md(
+        r#"<ol><li>item1</li><li>item2</li></ol>"#,
+        r#"1. item1
+2. item2"#,
+    );
+}
```

Comment:
I'd like to see some tests with multiple nested ordered and unordered lists, with formatting and code blocks inside them.

Reply:
OK!

Reply:
Done.